tools/state-viewer/src/dump_state_parts.rs

```diff
@@ -1,15 +1,96 @@
+use crate::epoch_info::iterate_and_filter;
+use clap::Subcommand;
 use near_chain::{ChainStore, ChainStoreAccess, RuntimeAdapter};
+use near_epoch_manager::EpochManager;
+use near_primitives::epoch_manager::epoch_info::EpochInfo;
 use near_primitives::state_part::PartId;
 use near_primitives::syncing::get_num_state_parts;
+use near_primitives::types::EpochId;
 use near_primitives_core::hash::CryptoHash;
-use near_primitives_core::types::ShardId;
+use near_primitives_core::types::{BlockHeight, EpochHeight, ShardId};
 use near_store::Store;
 use nearcore::{NearConfig, NightshadeRuntime};
 use std::path::Path;
+use std::str::FromStr;
 use std::sync::Arc;
 
+#[derive(Subcommand, Debug, Clone)]
+pub(crate) enum EpochSelection {
+    /// Current epoch.
+    Current,
+    /// Fetch the given epoch.
+    EpochId { epoch_id: String },
+    /// Fetch epochs at the given height.
+    EpochHeight { epoch_height: EpochHeight },
+    /// Fetch an epoch containing the given block hash.
+    BlockHash { block_hash: String },
+    /// Fetch an epoch containing the given block height.
+    BlockHeight { block_height: BlockHeight },
+}
+
+fn get_epoch_id(
+    epoch_selection: EpochSelection,
+    store: Store,
+    chain_store: &mut ChainStore,
+    epoch_manager: &mut EpochManager,
+) -> EpochId {
+    match epoch_selection {
+        EpochSelection::Current => {
+            epoch_manager.get_epoch_id(&chain_store.head().unwrap().last_block_hash).unwrap()
+        }
+        EpochSelection::EpochId { epoch_id } => EpochId(CryptoHash::from_str(&epoch_id).unwrap()),
+        EpochSelection::EpochHeight { epoch_height } => {
+            // Fetch epochs at the given height.
+            // There should only be one epoch at a given height. But this is a debug tool, let's check
+            // if there are multiple epochs at a given height.
+            let epoch_ids =
+                iterate_and_filter(store, |epoch_info| epoch_info.epoch_height() == epoch_height);
+            assert_eq!(epoch_ids.len(), 1, "{:#?}", epoch_ids);
+            epoch_ids[0].clone()
+        }
+        EpochSelection::BlockHash { block_hash } => {
+            let block_hash = CryptoHash::from_str(&block_hash).unwrap();
+            epoch_manager.get_epoch_id(&block_hash).unwrap()
+        }
+        EpochSelection::BlockHeight { block_height } => {
+            // Fetch an epoch containing the given block height.
+            let block_hash = chain_store.get_block_hash_by_height(block_height).unwrap();
+            epoch_manager.get_epoch_id(&block_hash).unwrap()
+        }
+    }
+}
+
+fn get_prev_hash_of_epoch(
+    epoch_info: &EpochInfo,
+    chain_store: &ChainStore,
+    epoch_manager: &mut EpochManager,
+) -> CryptoHash {
+    let head = chain_store.head().unwrap();
+    let mut cur_block_info = epoch_manager.get_block_info(&head.last_block_hash).unwrap();
+    loop {
```

Comment:
I'm surprised that you have to iterate over epochs like this. Could you just fetch the epoch from the epoch manager directly?

Reply:
Fetching an epoch is not enough. I need a block hash of the last block before the specified epoch.

I can fetch the epoch, but `EpochManager` has no functions mapping epochs to blocks.
`BlockInfo` has `epoch_first_block`, but I can't get any `BlockInfo` from `EpochInfo`.

Same is true for `RuntimeAdapter`. I could call `get_validator_info(EpochInfo)` to get `epoch_start_height`, but to get a block by height, I need an instance of `Chain`.

Reply:
Please let me know if I'm missing something obvious in the APIs.

Extending the API involves adding unit tests and faking the new methods. As that API is not needed for running the node, I prefer to workaround that.

Reply:
thanks for the explanation.